gateway/src/service.rs

```diff
@@ -591,35 +568,110 @@ pub mod tests {
         };
 
         svc.create_user(neo.clone()).await.unwrap();
+        svc.create_user(trinity.clone()).await.unwrap();
 
-        let work = svc
+        let project = svc
             .create_project(matrix.clone(), neo.clone())
             .await
             .unwrap();
 
-        // work work work work
-        let project = work.work;
-
         assert!(creating_same_project_name(&project, &matrix));
 
         assert_eq!(svc.find_project(&matrix).await.unwrap(), project);
 
-        let work = svc.destroy_project(matrix.clone(), neo).await.unwrap();
-
-        let project = work.work;
+        let mut work = svc
+            .new_task()
+            .project(matrix.clone())
+            .account(neo.clone())
+            .and_then(task::destroy())
+            .build();
 
-        assert!(matches!(&project, Project::Destroyed(_)));
-
-        svc.update_project(&matrix, &project).await.unwrap();
+        while let TaskResult::Pending(_) = work.poll(()).await {}
+        assert!(matches!(work.poll(()).await, TaskResult::Done(())));
 
+        // After project has been destroyed...
         assert!(matches!(
             svc.find_project(&matrix).await,
+            Ok(Project::Destroyed(_))
+        ));
+
+        // If recreated by a different user
+        assert!(matches!(
+            svc.create_project(matrix.clone(), trinity.clone()).await,
             Err(Error {
-                kind: ErrorKind::ProjectNotFound,
+                kind: ErrorKind::ProjectAlreadyExists,
                 ..
             })
         ));
 
+        // If recreated by the same user
+        assert!(matches!(
+            svc.create_project(matrix, neo).await,
+            Ok(Project::Creating(_))
+        ));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn service_create_ready_kill_restart_docker() -> anyhow::Result<()> {
+        let world = World::new().await;
+        let svc = Arc::new(GatewayService::init(world.args(), world.pool()).await);
+
+        let neo: AccountName = "neo".parse().unwrap();
+        let matrix: ProjectName = "matrix".parse().unwrap();
+
+        svc.create_user(neo.clone()).await.unwrap();
+        svc.create_project(matrix.clone(), neo.clone())
+            .await
+            .unwrap();
+
+        let mut task = svc
+            .new_task()
+            .account(neo.clone())
+            .project(matrix.clone())
+            .build();
+
+        while let TaskResult::Pending(_) = task.poll(()).await {
+            // keep polling
+        }
```

Comment:
Is there some build-in timeout to prevent this from polling indefinitely?

Reply:
Not in the tests no, that's a good point. There is a timeout for `Project` tasks but not generically for these. I'll quickly add that.